processor/src/main/java/org/mapstruct/ap/internal/processor/MapperCreationProcessor.java

```diff
@@ -446,41 +446,48 @@ private void mergeInheritedOptions(SourceMethod method, MapperConfiguration mapp
-            if ( applicablePrototypeMethods.size() == 1 ) {
-                mappingOptions.applyInheritedOptions(
-                    first( applicablePrototypeMethods ).getMappingOptions(),
-                    false,
-                    method,
-                    messager,
-                    typeFactory );
-            }
-            else if ( applicablePrototypeMethods.size() > 1 ) {
-                messager.printMessage(
-                    method.getExecutable(),
-                    Message.INHERITCONFIGURATION_MULTIPLE_PROTOTYPE_METHODS_MATCH,
-                    Strings.join( applicablePrototypeMethods, ", " ) );
+        else if ( inheritanceStrategy.isAutoInherit() ) {
+
+            if ( inheritanceStrategy.isApplyForward() ) {
+                if ( applicablePrototypeMethods.size() == 1 ) {
+                    mappingOptions.applyInheritedOptions(
+                        first( applicablePrototypeMethods ).getMappingOptions(),
+                        false,
+                        method,
+                        messager,
+                        typeFactory );
+                }
+                else if ( applicablePrototypeMethods.size() > 1 ) {
+                    messager.printMessage(
+                        method.getExecutable(),
+                        Message.INHERITCONFIGURATION_MULTIPLE_PROTOTYPE_METHODS_MATCH,
+                        Strings.join( applicablePrototypeMethods, ", " ) );
+                }
             }
 
-            if ( applicableReversePrototypeMethods.size() == 1 ) {
-                mappingOptions.applyInheritedOptions(
-                    first( applicableReversePrototypeMethods ).getMappingOptions(),
-                    true,
-                    method,
-                    messager,
-                    typeFactory );
-            }
-            else if ( applicableReversePrototypeMethods.size() > 1 ) {
-                messager.printMessage(
-                    method.getExecutable(),
-                    Message.INHERITINVERSECONFIGURATION_MULTIPLE_PROTOTYPE_METHODS_MATCH,
-                    Strings.join( applicablePrototypeMethods, ", " ) );
+            if ( inheritanceStrategy.isApplyReverse() ) {
+                if ( applicableReversePrototypeMethods.size() == 1 ) {
+                    mappingOptions.applyInheritedOptions(
+                        first( applicableReversePrototypeMethods ).getMappingOptions(),
+                        true,
+                        method,
+                        messager,
+                        typeFactory );
+                }
+                else if ( applicableReversePrototypeMethods.size() > 1 ) {
+                    messager.printMessage(
+                        method.getExecutable(),
+                        Message.INHERITINVERSECONFIGURATION_MULTIPLE_PROTOTYPE_METHODS_MATCH,
+                        Strings.join( applicablePrototypeMethods, ", " ) );
```

Comment:
This should be `applicableReversePrototypeMethods` instead of `applicablePrototypeMethods`. In any case, I think it would be better if we can extract it into a method. Something like:

```java
private void applyInheritedOptions(SourceMethod method, MappingOptions mappingOptions,
        List<SourceMethod> applicableMethods, boolean isInverse) {
        if ( applicableMethods.size() == 1 ) {
            mappingOptions.applyInheritedOptions(
                first( applicableMethods ).getMappingOptions(),
                isInverse,
                method,
                messager,
                typeFactory
            );
        }
        else if ( applicableMethods.size() > 1 ) {
            messager.printMessage(
                method.getExecutable(),
                isInverse ? Message.INHERITINVERSECONFIGURATION_MULTIPLE_PROTOTYPE_METHODS_MATCH :
                    Message.INHERITCONFIGURATION_MULTIPLE_PROTOTYPE_METHODS_MATCH,
                Strings.join( applicableMethods, ", " )
            );
        }
    }
```

And we just call it with the appropriate parameters in the ifs. The logic is the same with the only difference in the methods and the `isInverse`